src/Lean/Server/FileWorker/RequestHandling.lean

```diff
@@ -134,6 +134,102 @@ partial def handleDefinition (kind : GoToKind) (p : TextDocumentPositionParams)
             return ← ci.runMetaM i.lctx <| locationLinksFromDecl i ei.elaborator
       return #[]
 
+inductive RefIdent where
+  | name : Name → RefIdent
+  | fvar : FVarId → RefIdent
+  deriving BEq
+
+structure Reference where
+  ident : RefIdent
+  range : Range
+  isDeclaration : Bool
+  deriving BEq
+
+open Std in
+open Elab in
+partial def handleReferences (p : ReferenceParams)
+    : RequestM (RequestTask (Array Location)) := do
+  let doc ← readDoc
+  let text := doc.meta.text
+  let hoverPos := text.lspPosToUtf8Pos p.position
+  let t ← doc.cmdSnaps.waitAll
+  mapTask t fun (snaps, _) => do
+    if let some snap := snaps.find? (fun s => s.endPos > hoverPos) then
+      if let some (ci, i) := snap.infoTree.hoverableInfoAt? hoverPos then
+        if let some (ident, _) := identOf i then
+          return referencesTo doc ident (p.context.includeDeclaration) snaps
+    return #[]
+  where
+    identOf : Info → Option (RefIdent × Bool)
+      | Info.ofTermInfo ti => match ti.expr with
+        | Expr.const n .. => some (RefIdent.name n, ti.isBinder)
+        | Expr.fvar id .. => some (RefIdent.fvar id, ti.isBinder)
+        | _ => none
+      | Info.ofFieldInfo fi => some (RefIdent.name fi.projName, false)
+      | _ => none
+
+    addReference (doc : EditableDocument) (info : Info)
+        (ident : RefIdent) (isDeclaration : Bool)
+        (refs : Array Reference) : Array Reference :=
+      if let some range := info.stx.getRange? (originalOnly := true) then
+        let text := doc.meta.text
+        let range := {
+          start := text.utf8PosToLspPos range.start
+          «end» := text.utf8PosToLspPos range.stop
+        }
+        refs.push { ident, range, isDeclaration }
+      else
+        refs
+
+    addReferences (doc : EditableDocument) (snap : Snapshot)
+        (refs : Array Reference) : Array Reference :=
+      let infos := snap.infoTree.findAll fun
+        | Info.ofTermInfo _ => true
+        | Info.ofFieldInfo _ => true
+        | _ => false
+      infos.foldl (init := refs) fun refs info => do
+        if let some (ident, isDecl) := identOf info then
+          return addReference doc info ident isDecl refs
+        return refs
+
+    findReferences (doc : EditableDocument) (snaps : List Snapshot) : Array Reference :=
+      snaps.foldl (init := #[]) fun refs snap => addReferences doc snap refs
```

Comment:
You might consider inlining `addReferences` and `addReference` into a big loop(s) with a mutable variable to get rid of the many parameters, but this way is also fine